### PR TITLE
feat: add IP and origin filtering for trusted gateways

### PR DESCRIPTION
**Pull request by djwhitt**
## Summary

This PR implements comprehensive IP and origin filtering for trusted gateway forwarding to prevent loops and enhance security. The implementation has been significantly enhanced based on reviewer feedback to address security vulnerabilities and improve robustness.

### 🛡️ Enhanced Security
- **X-Forwarded-For spoofing prevention**: Always includes `socket.remoteAddress` even when X-Forwarded-For is present
- **Robust IP validation**: Validates IPv4 octets (0-255) and rejects invalid formats
- **IPv4-mapped IPv6 normalization**: Handles `::ffff:192.168.1.1` → `192.168.1.1`
- **Leading zero protection**: Prevents octal interpretation attacks (`192.168.001.1` rejected)
- **Comprehensive edge case handling**: Validates CIDR prefixes, handles /0 and /32 masks

### 🔧 Flexible Configuration
- **Dual IP/CIDR support**: Individual IPs (IPv4/IPv6) and IPv4 CIDR ranges
- **Comma-separated format**: Simple configuration syntax
- **Bare IP support**: Treats individual IPs as /32 CIDRs automatically

### 📊 Comprehensive Testing
- **31 test cases** covering all edge cases and security scenarios
- **IPv6 support validation**: Exact matching for individual IPv6 addresses
- **CIDR validation**: Invalid prefixes (/33, /-1), edge cases (/0, /32)
- **Mixed format testing**: IPv4 + IPv6 + CIDR in same configuration

### Environment Variable
```bash
# Individual IPs and CIDR ranges (comma-separated)
TRUSTED_GATEWAYS_BLOCKED_IPS_AND_CIDRS="192.168.1.1,10.0.0.0/8,2001:db8::1,172.16.0.0/12"
```

### Code Usage
```typescript
const filteredSource = new FilteredContiguousDataSource({
  log,
  dataSource: innerSource,
  blockedOrigins: ['evil.gateway.com'],
  blockedIpsAndCidrs: ['192.168.1.1', '10.0.0.0/8', '2001:db8::1']
});
```

### IP Collection & Validation
- Processes both string and string[] X-Forwarded-For headers
- Filters out "unknown" and invalid entries
- Normalizes IPv4-mapped IPv6 addresses at source
- Validates IP formats before adding to collection
- Preserves order: X-Forwarded-For IPs first, then remoteAddress

### CIDR Matching Enhancement
- **Fixed security bugs**: Proper IPv4 octet validation (0-255)
- **Bare IP support**: Automatically treats `192.168.1.1` as `192.168.1.1/32`
- **Edge case handling**: Correct /0 mask calculation, invalid prefix rejection
- **IPv6 exact matching**: String-based matching for individual IPv6 addresses

### Interface Improvements
- **Exported RequestAttributes**: Now available for external imports
- **Enhanced JSDoc**: Comprehensive documentation for all interfaces
- **Parameter clarity**: `blockedIpsAndCidrs` accurately reflects functionality

### Prevents Common Attack Vectors
1. **Header spoofing**: X-Forwarded-For manipulation can't bypass remoteAddress inclusion
2. **Invalid IP bypass**: Malformed IPs (999.999.999.999) don't match valid entries  
3. **IPv6 bypass**: IPv4-mapped IPv6 addresses properly normalized
4. **Octal interpretation**: Leading zeros rejected to prevent parsing confusion
5. **CIDR edge cases**: Invalid prefixes and edge masks handled securely

### Backward Compatibility
- Maintains `clientIp` field for legacy compatibility
- Environment variable rename with clear migration path
- Existing functionality preserved while adding new capabilities

## Files Changed
- `src/data/filtered-contiguous-data-source.ts` - Core filtering logic and security fixes
- `src/routes/data/handlers.ts` - Enhanced X-Forwarded-For parsing with validation
- `src/config.ts` - Environment variable rename and validation updates
- `src/system.ts` - Configuration integration
- `src/types.d.ts` - Exported interface with JSDoc documentation
- `.env.example` - Updated configuration example
- `docs/envs.md` - Documentation updates
- Test files - Comprehensive edge case coverage (31 tests)

## Testing

All tests passing: **31 test cases** including:
- Basic origin and IP filtering functionality
- IPv6 exact matching scenarios
- IPv4-mapped IPv6 normalization
- Invalid CIDR and IP format handling
- Edge cases (/0, /32 masks, bare IPs)
- Security scenarios (leading zeros, mixed formats)
- X-Forwarded-For header parsing validation

This implementation provides robust protection against gateway loops while supporting flexible IP and CIDR-based filtering with comprehensive security validations.

### CODE REVIEW

File: src/routes/data/handlers.ts [hunk 1]
```diff
@@ -203,13 +203,27 @@ export const getRequestAttributes = (
     originNodeRelease = nodeRelease;
   }
 
+  // Extract client IP from request headers (X-Forwarded-For) or connection
+  let clientIp: string | undefined;
+  const xForwardedFor = req.headers['x-forwarded-for'] as string;
```
Comment (djwhitt) [inline src/routes/data/handlers.ts]: Can we handle the types better here? Perhaps `unknown` => `string`.

File: src/config.ts [hunk 2]
```diff
@@ -135,6 +135,35 @@ Object.entries(TRUSTED_GATEWAYS_URLS).forEach(([url, weight]) => {
...
+});
+
+// Trusted gateways blocked IP addresses (IPs/CIDR ranges to reject when forwarding)
+export const TRUSTED_GATEWAYS_BLOCKED_IP_ADDRESSES = JSON.parse(
+  env.varOrDefault('TRUSTED_GATEWAYS_BLOCKED_IP_ADDRESSES', '[]'),
+) as string[];
+
+// Basic validation for IP addresses - just check they're strings for now
+// Could add more sophisticated IP/CIDR validation if needed
+TRUSTED_GATEWAYS_BLOCKED_IP_ADDRESSES.forEach((ip) => {
```
Comment (djwhitt) [inline src/config.ts]: Let's make this just CIDRs instead of IPs and CIDRs.

File: .env.example [hunk 3]
```diff
@@ -10,6 +10,15 @@
 # Arweave node to use for proxying requests
 # TRUSTED_GATEWAY_URL="https://arweave.net"
 
+# Trusted gateways configuration (JSON object with URL keys and priority values)
+# TRUSTED_GATEWAYS_URLS='{"https://arweave.net": 1, "https://ar-io.dev": 2}'
+
+# Origins to block when forwarding to trusted gateways (JSON array)
+# TRUSTED_GATEWAYS_BLOCKED_ORIGINS='["self.gateway.com", "loop.source.net"]'
```
Comment (djwhitt) [inline .env.example]: Let's make this a comma separated list.

File: .env.example [hunk 4]
```diff
@@ -10,6 +10,15 @@
...
+
+# IP addresses/CIDR ranges to block when forwarding to trusted gateways (JSON array)
+# TRUSTED_GATEWAYS_BLOCKED_IP_ADDRESSES='["192.168.1.100", "10.0.0.0/8"]'
```
Comment (djwhitt) [inline .env.example]: Let's make this a comma separated list.

File: src/routes/data/handlers.test.ts [hunk 5]
```diff
@@ -1517,7 +1530,7 @@ st
             .expect(200)
             .then((res: any) => {
               assert.equal(res.headers['x-ar-io-trusted'], 'false');
-              assert.equal(res.headers['x-ar-io-cache'], 'MISS');
+              assert.equal(res.headers['x-cache'], 'MISS');
               assert.equal(res.body.toString(), 'testing...');
             });
```
[the remark this answers is not in the document]
Comment (djwhitt) [reply src/routes/data/handlers.test.ts:1535]: @coderabbitai I think this has been fixed now. Can you confirm?

File: src/data/filtered-contiguous-data-source.ts [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+    if (requestAttributes) {
+      if (this.isOriginBlocked(requestAttributes.origin)) {
+        this.log.info('Blocking request due to blocked origin', {
+          id,
+          origin: requestAttributes.origin,
+        });
+        throw new Error(
+          `Request blocked: origin '${requestAttributes.origin}' is blocked`,
+        );
+      }
```
[the remark this answers is not in the document]
Comment (djwhitt) [reply src/data/filtered-contiguous-data-source.ts:140]: @coderabbitai I'm not worried about this.

File: src/data/filtered-contiguous-data-source.ts [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+            cidr: blockedCidr,
+            allIps: ipsToCheck,
+          });
+          return true;
+        }
+      }
+    }
+
+    return false;
+  }
```
[the remark this answers is not in the document]
Comment (djwhitt) [reply src/data/filtered-contiguous-data-source.ts:87]: @coderabbitai Please review the committed solution and tell me what you think.

File: src/data/filtered-contiguous-data-source.test.ts [hunk 8]
```diff
@@ -0,0 +1,426 @@
...
+  });
+
+  describe('Configuration', () => {
+    it('should work with empty blocking lists', async () => {
+      const unrestrictedDataSource = new FilteredContiguousDataSource({
+        log,
+        dataSource: mockDataSource,
+        blockedOrigins: [],
+        blockedIpAddresses: [],
+      });
```
[the remark this answers is not in the document]
Comment (djwhitt) [reply src/data/filtered-contiguous-data-source.test.ts:258]: @coderabbitai Please review the committed solution and tell me what you think.

File: src/data/filtered-contiguous-data-source.test.ts [hunk 9]
```diff
@@ -0,0 +1,426 @@
...
+
+    it('should handle invalid CIDR gracefully', async () => {
+      const filteredDataSourceWithInvalidCidr =
+        new FilteredContiguousDataSource({
+          log,
+          dataSource: mockDataSource,
+          blockedOrigins: [],
+          blockedIpAddresses: ['invalid-cidr/abc'],
+        });
+
```
[the remark this answers is not in the document]
Comment (djwhitt) [reply src/data/filtered-contiguous-data-source.test.ts:232]: @coderabbitai Please review the committed solution and tell me what you think.

**Comment by djwhitt** [comment]
@coderabbitai Please review again and note whether the issues you mentioned above

**Comment by djwhitt** [comment]
@coderabbitai Please review again and note whether the issues you mentioned above have been fixed.
